### DISCUSSION TITLE
How can I load legacy.dll in an MFC application?

### DISCUSSION THREAD

**Discussion by so1616** [openssl/openssl#24852] [Q&A] [answered]
I'm using OpenSSL version 3.0.14.
When calling the dll that loads legacy.dll with OSSL_PROVIDER_load from an x86 MFC application, legacy.dll fails to load.
When calling the dll from an x86 console application, legacy.dll is successfully loaded.
How do I load legacy.dll in an MFC application?

MFC application
```
BOOL TestDlg::OnInitDialog()
{
	typedef int (*dll_Test)(void);
	
	HMODULE hModule = LoadLibraryW(L"C:\\Users\\...........\\OpenSSLTestDll.dll");

	if (!hModule) {
		return false;
	}

	dll_Test func1 = (dll_Test)GetProcAddress(hModule, "Test");
	if (func1 != NULL) {
		func1();
	}
	return true;
}
```

OpenSSLTestDll.dll
```
#include "pch.h"
#include "openssl/evp.h"
#include "openssl/provider.h"

extern "C" __declspec(dllexport)int Test()
{
	typedef OSSL_PROVIDER* (*dll_OSSL_PROVIDER_load)(OSSL_LIB_CTX*, const char*);
	typedef int (*dll_OSSL_PROVIDER_unload)(OSSL_PROVIDER*);

	HMODULE hModule = LoadLibraryW(L"C:\\test\\libcrypto-3.dll");
	if (!hModule) {
		return 0;
	}

	dll_OSSL_PROVIDER_load lp_dll_OSSL_PROVIDER_load = (dll_OSSL_PROVIDER_load)GetProcAddress(hModule, "OSSL_PROVIDER_load");
	dll_OSSL_PROVIDER_unload lp_dll_OSSL_PROVIDER_unload = (dll_OSSL_PROVIDER_unload)GetProcAddress(hModule, "OSSL_PROVIDER_unload");

	OSSL_PROVIDER* legacy = lp_dll_OSSL_PROVIDER_load(NULL, "legacy");
	if (legacy == NULL) {
		return 0;
	}else{
		lp_dll_OSSL_PROVIDER_unload(legacy);
		return 1;
	}
}

```

**Comment by mattcaswell** [comment, marked as answer]
When you call `OSSL_PROVIDER_load(NULL, "legacy")` this will cause OpenSSL to look for the file `legacy.dll`. If this call fails it is usually because it cannot find that dll.

If the `OPENSSL_MODULES` environment variable has been set then OpenSSL will look in the folder that that env variable points at to attempt to load `legacy.dll`. If the environment variable is not set then it will look in `MODULESDIR` - which is a location determined at OpenSSL compile time. You can check the location of `MODULESDIR` using the OpenSSL command line app:

````
openssl version -m
````
Alternatively you can check that location by calling `OpenSSL_version(OPENSSL_MODULES_DIR);` from your code and looking at the return value which will be in the form "MODULESDIR: ..."

Check the folder where OpenSSL is looking for the dll and confirm that it is present.

**Reply by so1616** [reply]
Thank you for the reply.

I would like to specify the path to load legacy.dll.
However, loading from an MFC application failed, just as when path was not specified.
Loading from console application was successful.


OpenSSLTestDll.dll

```
#include "pch.h"
#include "openssl/evp.h"
#include "openssl/provider.h"

extern "C" __declspec(dllexport)int Test()
{
	typedef OSSL_PROVIDER* (*dll_OSSL_PROVIDER_load)(OSSL_LIB_CTX*, const char*);
	typedef int (*dll_OSSL_PROVIDER_unload)(OSSL_PROVIDER*);
	
	/* <- add */
	typedef OSSL_LIB_CTX* (*dll_OSSL_LIB_CTX_new)(void);
	typedef void (*dll_OSSL_LIB_CTX_free)(OSSL_LIB_CTX*);
	typedef OSSL_LIB_CTX* (*dll_OSSL_LIB_CTX_get0_global_default)(void);
	typedef int (*dll_OSSL_PROVIDER_set_default_search_path)(OSSL_LIB_CTX*, const char*);
	/* add -> */

	HMODULE hModule = LoadLibraryW(L"C:\\test\\libcrypto-3.dll");
	if (!hModule) {
		return 0;
	}

	dll_OSSL_PROVIDER_load lp_dll_OSSL_PROVIDER_load = (dll_OSSL_PROVIDER_load)GetProcAddress(hModule, "OSSL_PROVIDER_load");
	dll_OSSL_PROVIDER_unload lp_dll_OSSL_PROVIDER_unload = (dll_OSSL_PROVIDER_unload)GetProcAddress(hModule, "OSSL_PROVIDER_unload");
	
	/* <- add */
	dll_OSSL_LIB_CTX_new lp_dll_OSSL_LIB_CTX_new = (dll_OSSL_LIB_CTX_new)GetProcAddress(hModule, "OSSL_LIB_CTX_new");
	dll_OSSL_LIB_CTX_free lp_dll_OSSL_LIB_CTX_free = (dll_OSSL_LIB_CTX_free)GetProcAddress(hModule, "OSSL_LIB_CTX_free");
	dll_OSSL_LIB_CTX_get0_global_default lp_dll_OSSL_LIB_CTX_get0_global_default = (dll_OSSL_LIB_CTX_get0_global_default)GetProcAddress(hModule, "OSSL_LIB_CTX_get0_global_default");
	dll_OSSL_PROVIDER_set_default_search_path lp_dll_OSSL_PROVIDER_set_default_search_path = (dll_OSSL_PROVIDER_set_default_search_path)GetProcAddress(hModule, "OSSL_PROVIDER_set_default_search_path");
	
	OSSL_LIB_CTX* osslLibCtx = lp_dll_OSSL_LIB_CTX_new();
	osslLibCtx = lp_dll_OSSL_LIB_CTX_get0_global_default();
	lp_dll_OSSL_PROVIDER_set_default_search_path(osslLibCtx, "C:\\Program Files\\Test\\lib\\");
	/* add -> */
	
	OSSL_PROVIDER* legacy = lp_dll_OSSL_PROVIDER_load(NULL, "legacy");
	if (legacy == NULL) {
		lp_dll_OSSL_LIB_CTX_free(osslLibCtx);
		return 0;
	}else{
		lp_dll_OSSL_PROVIDER_unload(legacy);
		lp_dll_OSSL_LIB_CTX_free(osslLibCtx);
		return 1;
	}
}
```

**Reply by mattcaswell** [reply]
````
	OSSL_LIB_CTX* osslLibCtx = lp_dll_OSSL_LIB_CTX_new();
	osslLibCtx = lp_dll_OSSL_LIB_CTX_get0_global_default();
````
This part seems unnecessary. Just passing `NULL` as the first parameter to `OSSL_PROVIDER_set_default_search_path` should set the search path for the default libctx.

````
	lp_dll_OSSL_PROVIDER_set_default_search_path(osslLibCtx, "C:\\Program Files\\Test\\lib\\");
````
Stupid question, but I assume legacy.dll does actually exist in that location?

Do you have an OpenSSL config file that attempts to load any providers?

After the failed call I suggest writing out the OpenSSL error stack somewhere, e.g. something like this

````
BIO *errs = BIO_new_file("error.log", "w");
ERR_print_errors(errs);
````

**Reply by so1616** [reply]
Thank you for your advice.
legacy.dll has been successfully loaded.
It was a mistake to not be able to load legacy.dll in an MFC application.
I'm sorry.

I would like to rename and use libcrypto-3.dll.
However, when I rename it, legacy.dll will fail to load.
Is it possible to rename libcrypto-3.dll and use it?

**Reply by mattcaswell** [reply]
Hmmm. I'm actually not sure. Perhaps @levitte might comment on this.

**Reply by t8m** [reply]
You would have to create your own configuration target with `shlib_variant` setting. Please see https://github.com/openssl/openssl/blob/master/INSTALL.md#creating-your-own-configuration

The file would contain something like:
```
my %targets = (
    "VC-WIN64A-3" => {
        inherit_from     => [ "VC-WIN64A" ],
        shlib_variant    => "-3"
    }
}
```

Then you would call Configure VC-WIN64A-3

**Reply by so1616** [reply]
I want to rename it libcrypto_3.dll.
When using this conf file, it was renamed to libcrypto-3_3.dll.
Is -3 always included?

99-ourtargets.conf
```
(
    "VC-WIN32_3" => {
        inherit_from => [ 'VC-WIN32' ],
        shlib_variant => "_3",
    }
);
```

**Reply by t8m** [reply]
Yes, it is.

**Reply by so1616** [reply]
@mattcaswell 
@t8m 

Thank you for your response.
All my problems were solved.
Thank you very much.
